### PR TITLE
ISSUE 145 - Implement auto-deployment

### PR DESCRIPTION
**Pull request by prayagupd**
uses [sbt-release](https://github.com/sbt/sbt-release) plugin to solve https://github.com/scalaz/scalaz-zio/issues/145

- merge on master should trigger `sbt "release with-defaults"` which will 

1) create a tag with name `v${current.version}-commitSha`
2) publish artifact with versioning `${current.version}-commitSha`
3) push a tag

I tested in local with `mavenLocal` as publish repo, and I can see following 4 artifacts created for zio, zio-interop, scalajs, scalajs-interop

```bash
$ ls -l ~/.m2/repository/org/scalaz/scalaz-zio_2.12/0.1.0-3e8941e/
total 3536
-rw-r--r--  1 prayagupd  NA\Domain Users  1515611 Jul 31 20:35 scalaz-zio_2.12-0.1.0-3e8941e-javadoc.jar
-rw-r--r--  1 prayagupd  NA\Domain Users       32 Jul 31 20:35 scalaz-zio_2.12-0.1.0-3e8941e-javadoc.jar.md5
-rw-r--r--  1 prayagupd  NA\Domain Users       40 Jul 31 20:35 scalaz-zio_2.12-0.1.0-3e8941e-javadoc.jar.sha1
-rw-r--r--  1 prayagupd  NA\Domain Users    31908 Jul 31 20:35 scalaz-zio_2.12-0.1.0-3e8941e-sources.jar
-rw-r--r--  1 prayagupd  NA\Domain Users       32 Jul 31 20:35 scalaz-zio_2.12-0.1.0-3e8941e-sources.jar.md5
-rw-r--r--  1 prayagupd  NA\Domain Users       40 Jul 31 20:35 scalaz-zio_2.12-0.1.0-3e8941e-sources.jar.sha1
-rw-r--r--  1 prayagupd  NA\Domain Users   217682 Jul 31 20:35 scalaz-zio_2.12-0.1.0-3e8941e.jar
-rw-r--r--  1 prayagupd  NA\Domain Users       32 Jul 31 20:35 scalaz-zio_2.12-0.1.0-3e8941e.jar.md5
-rw-r--r--  1 prayagupd  NA\Domain Users       40 Jul 31 20:35 scalaz-zio_2.12-0.1.0-3e8941e.jar.sha1
-rw-r--r--  1 prayagupd  NA\Domain Users     2129 Jul 31 20:35 scalaz-zio_2.12-0.1.0-3e8941e.pom
-rw-r--r--  1 prayagupd  NA\Domain Users       32 Jul 31 20:35 scalaz-zio_2.12-0.1.0-3e8941e.pom.md5
-rw-r--r--  1 prayagupd  NA\Domain Users       40 Jul 31 20:35 scalaz-zio_2.12-0.1.0-3e8941e.pom.sha1
```

```bash
$ ls -l ~/.m2/repository/org/scalaz/scalaz-zio-interop_2.12/0.1.0-3e8941e/
total 2504
-rw-r--r--  1 prayagupd  NA\Domain Users  1173093 Jul 31 20:35 scalaz-zio-interop_2.12-0.1.0-3e8941e-javadoc.jar
-rw-r--r--  1 prayagupd  NA\Domain Users       32 Jul 31 20:35 scalaz-zio-interop_2.12-0.1.0-3e8941e-javadoc.jar.md5
-rw-r--r--  1 prayagupd  NA\Domain Users       40 Jul 31 20:35 scalaz-zio-interop_2.12-0.1.0-3e8941e-javadoc.jar.sha1
-rw-r--r--  1 prayagupd  NA\Domain Users     3238 Jul 31 20:35 scalaz-zio-interop_2.12-0.1.0-3e8941e-sources.jar
-rw-r--r--  1 prayagupd  NA\Domain Users       32 Jul 31 20:35 scalaz-zio-interop_2.12-0.1.0-3e8941e-sources.jar.md5
-rw-r--r--  1 prayagupd  NA\Domain Users       40 Jul 31 20:35 scalaz-zio-interop_2.12-0.1.0-3e8941e-sources.jar.sha1
-rw-r--r--  1 prayagupd  NA\Domain Users    64245 Jul 31 20:35 scalaz-zio-interop_2.12-0.1.0-3e8941e.jar
-rw-r--r--  1 prayagupd  NA\Domain Users       32 Jul 31 20:35 scalaz-zio-interop_2.12-0.1.0-3e8941e.jar.md5
-rw-r--r--  1 prayagupd  NA\Domain Users       40 Jul 31 20:35 scalaz-zio-interop_2.12-0.1.0-3e8941e.jar.sha1
-rw-r--r--  1 prayagupd  NA\Domain Users     2561 Jul 31 20:35 scalaz-zio-interop_2.12-0.1.0-3e8941e.pom
-rw-r--r--  1 prayagupd  NA\Domain Users       32 Jul 31 20:35 scalaz-zio-interop_2.12-0.1.0-3e8941e.pom.md5
-rw-r--r--  1 prayagupd  NA\Domain Users       40 Jul 31 20:35 scalaz-zio-interop_2.12-0.1.0-3e8941e.pom.sha1
```
scalajs

```bash
$ ls -l ~/.m2/repository/org/scalaz/scalaz-zio_sjs0.6_2.12/0.1.0-3e8941e/
total 3592
-rw-r--r--  1 prayagupd  NA\Domain Users  1485040 Jul 31 20:35 scalaz-zio_sjs0.6_2.12-0.1.0-3e8941e-javadoc.jar
-rw-r--r--  1 prayagupd  NA\Domain Users       32 Jul 31 20:35 scalaz-zio_sjs0.6_2.12-0.1.0-3e8941e-javadoc.jar.md5
-rw-r--r--  1 prayagupd  NA\Domain Users       40 Jul 31 20:35 scalaz-zio_sjs0.6_2.12-0.1.0-3e8941e-javadoc.jar.sha1
-rw-r--r--  1 prayagupd  NA\Domain Users    22174 Jul 31 20:35 scalaz-zio_sjs0.6_2.12-0.1.0-3e8941e-sources.jar
-rw-r--r--  1 prayagupd  NA\Domain Users       32 Jul 31 20:35 scalaz-zio_sjs0.6_2.12-0.1.0-3e8941e-sources.jar.md5
-rw-r--r--  1 prayagupd  NA\Domain Users       40 Jul 31 20:35 scalaz-zio_sjs0.6_2.12-0.1.0-3e8941e-sources.jar.sha1
-rw-r--r--  1 prayagupd  NA\Domain Users   289009 Jul 31 20:35 scalaz-zio_sjs0.6_2.12-0.1.0-3e8941e.jar
-rw-r--r--  1 prayagupd  NA\Domain Users       32 Jul 31 20:35 scalaz-zio_sjs0.6_2.12-0.1.0-3e8941e.jar.md5
-rw-r--r--  1 prayagupd  NA\Domain Users       40 Jul 31 20:35 scalaz-zio_sjs0.6_2.12-0.1.0-3e8941e.jar.sha1
-rw-r--r--  1 prayagupd  NA\Domain Users     2562 Jul 31 20:35 scalaz-zio_sjs0.6_2.12-0.1.0-3e8941e.pom
-rw-r--r--  1 prayagupd  NA\Domain Users       32 Jul 31 20:35 scalaz-zio_sjs0.6_2.12-0.1.0-3e8941e.pom.md5
-rw-r--r--  1 prayagupd  NA\Domain Users       40 Jul 31 20:35 scalaz-zio_sjs0.6_2.12-0.1.0-3e8941e.pom.sha1
```

```bash
$ ls -l ~/.m2/repository/org/scalaz/scalaz-zio-interop_sjs0.6_2.12/0.1.0-3e8941e/
total 2544
-rw-r--r--  1 prayagupd  NA\Domain Users  1167843 Jul 31 20:35 scalaz-zio-interop_sjs0.6_2.12-0.1.0-3e8941e-javadoc.jar
-rw-r--r--  1 prayagupd  NA\Domain Users       32 Jul 31 20:35 scalaz-zio-interop_sjs0.6_2.12-0.1.0-3e8941e-javadoc.jar.md5
-rw-r--r--  1 prayagupd  NA\Domain Users       40 Jul 31 20:35 scalaz-zio-interop_sjs0.6_2.12-0.1.0-3e8941e-javadoc.jar.sha1
-rw-r--r--  1 prayagupd  NA\Domain Users     2450 Jul 31 20:35 scalaz-zio-interop_sjs0.6_2.12-0.1.0-3e8941e-sources.jar
-rw-r--r--  1 prayagupd  NA\Domain Users       32 Jul 31 20:35 scalaz-zio-interop_sjs0.6_2.12-0.1.0-3e8941e-sources.jar.md5
-rw-r--r--  1 prayagupd  NA\Domain Users       40 Jul 31 20:35 scalaz-zio-interop_sjs0.6_2.12-0.1.0-3e8941e-sources.jar.sha1
-rw-r--r--  1 prayagupd  NA\Domain Users    88996 Jul 31 20:35 scalaz-zio-interop_sjs0.6_2.12-0.1.0-3e8941e.jar
-rw-r--r--  1 prayagupd  NA\Domain Users       32 Jul 31 20:35 scalaz-zio-interop_sjs0.6_2.12-0.1.0-3e8941e.jar.md5
-rw-r--r--  1 prayagupd  NA\Domain Users       40 Jul 31 20:35 scalaz-zio-interop_sjs0.6_2.12-0.1.0-3e8941e.jar.sha1
-rw-r--r--  1 prayagupd  NA\Domain Users     3008 Jul 31 20:35 scalaz-zio-interop_sjs0.6_2.12-0.1.0-3e8941e.pom
-rw-r--r--  1 prayagupd  NA\Domain Users       32 Jul 31 20:35 scalaz-zio-interop_sjs0.6_2.12-0.1.0-3e8941e.pom.md5
-rw-r--r--  1 prayagupd  NA\Domain Users       40 Jul 31 20:35 scalaz-zio-interop_sjs0.6_2.12-0.1.0-3e8941e.pom.sha1
```

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,3 +1,4 @@
 target
 .sbtopts
 project/.sbt
+.idea/
```
Comment (ktonga) [inline .gitignore]: Revert this please

File: build.sbt [hunk 2]
```diff
@@ -154,3 +155,19 @@ lazy val microsite = project.module
       "white-color"     -> "#FFFFFF"
     )
   )
+
+lazy val commitSha = Process("git rev-parse --short HEAD").lineStream.head
+
+releaseVersion := ((version: String) => s"${version}-${commitSha}" )
+releaseTagName := s"v${version.value}"
```
Comment (ktonga) [inline build.sbt]: The tag will be redundant with the merge commit

File: build.sbt [hunk 3]
```diff
@@ -154,3 +155,19 @@ lazy val microsite = project.module
...
+
+releaseProcess := Seq[ReleaseStep](
+  checkSnapshotDependencies,
+  inquireVersions,
+  runClean,
+  runTest,
+  setReleaseVersion,
```
Comment (ktonga) [inline build.sbt]: We don't want these: `setReleaseVersion`, `tagRelease`, `pushChanges`

**Comment by ktonga** [comment]
TBH I think we don't need release-plugin for this.
Something like [sbt-dynver](https://github.com/dwijnand/sbt-dynver) should be enough and preferred IMO.
SZ8 is using it if you're looking for inspiration https://github.com/scalaz/scalaz/blob/series/8.0.x/build.sbt#L15

File: build.sbt (hunk 3 above)
Comment (prayagupd) [reply build.sbt]: - sbt-release builds desired artifact based on new version set on `version.sbt`, which is set by `setReleaseVersion`. So, we want `setReleaseVersion`. but I should look into reverting the `version.sbt` to original version once build is successful.

- `tagRelease` is used to create a tag for each build, but depends what responsibility we want with `sbt` and what we want manual. I had a comment to do the same [using bash script](https://github.com/scalaz/scalaz-zio/issues/145#issuecomment-408970343) as I can create tag from scripting as well. 

I would be fine with `sbt-release` creating tags, unless contributors have permissions to push their dev tags (with `git push`) which will cause problems.

- I agree we do not need `pushChanges` and don't want to keep pushing dev tags. which means we need to either introduce some sort of flag which will let push the tags from CI/CD.
or go write scrip and simply `git push --tags`

File: .gitignore (hunk 1 above)
Comment (prayagupd) [reply .gitignore]: Hmm, I thought that should be fine. It might have to be ignored one day anyway. **I reverted the change**

Comment (ktonga) [reply .gitignore]: I added a note in master with the reason it shouldn't be added

**Comment by jdegoes** [comment]
@prayagupd Thanks! The build is failing.

**Comment by prayagupd** [comment]
Hi @ktonga I looked into [sbt-DynVer](https://github.com/dwijnand/sbt-dynver). It looks good but creates versioning based on git latest tag - https://github.com/dwijnand/sbt-dynver#detail

I don't know if that's the practice being followed in scalaz8, because i personally found it bit weird to have version like `0.0.1-e499f56+0-e499f568+20180805-1315.jar`, which is the default behaviour of sbt-DynVer.

We HAVE TO set custom versioning from what sbt-DynVer results the version to be. Their version is of format `${last_tag_version}+{number_of_new_commits}-latestCommitSha`. 
And our expected version is `$version-$commitSha`. And we also want to be able to publish other than dev builds like RCx or Mx, which sbt-DynVer is not good for as far I can tell while we can tell the version to release using `sbt-release`

Let me know what you think? I can look into re-writing version using `sbt-DynVer` if you think thats better. To me `sbt-release` looks fine.

@jdegoes I fix the build. apparently did not do proper formatting :)

**Comment by ktonga** [comment]
My opinion is that we want to do two totally different things here.

1. Auto-deploy on merge to master a development (unstable) version of the lib but we still want it to be reproducible and immutable, move away from snapshots
2. From time to time cut a release and deploy to maven central which should become the official distribution channel for GP and stable releases.

I think this PR is mostly about addressing 1. and I think dyn-ver is a perfect fit for it. Also I thing we should never invoke sbt release from Travis, we should stick to publish. We don't have to use the default from dyn-ver, we can come up with a nice and useful format of our own.

Regarding 2. I think for that we will need to introduce sbt-release and perform taggin' and shit but it should be invoked manually by one of the ZIO's maintainers who will be able to set a version to something meaningful during the process, like cut a few milestones then some release candidates and finally a release.

**Comment by ktonga** [comment]
Please consider also including #155 as part of the outcome of this change, including the SHA in the version is useful but makes picking the right one a bit more complicated.

**Comment by jdegoes** [comment]
@ktonga This PR still allows (2), right?

**Comment by ktonga** [comment]
Yeah sure, in that case we can keep sbt-release with the default steps (including tagging and pushing) but don't touch travis and make dynver work for publish

**Comment by prayagupd** [comment]
@ktonga I was trying to kill both with `sbt-release` :) . Here's what this PR does

1) for dev versions, `sbt "release with-defaults"` will create `$current_dev_version-$last_commitSha` artifacts. 
2) for release versions, manually have to fire `sbt "release release-version $current_release_version next-release-version current_version+1"`

As you said we want to stick with `sbt publish` for development builds, I honestly don't see much value in using sbt-DynVer, because sbt-DynVer takes versioning based on previous tag. And if we have to re-write the version what sbt-DynVer generates, then why even use it? 

But bigger problem I think is sbt-release and sbt-DynVer don't work very well if we want unique artifact for each build. Because sbt-release expects `version.sbt` should have `version in ThisBuild` stuff that way we can increment once we do the release. - https://github.com/sbt/sbt-release#release-process

But if there is `version.sbt` or any `version in ThisBuild` defined in sbt, `sbt-DynVer` will override with the same and we loose what we are expect the sbt-DynVer to do.

**As per your feedback above,**
1) I will look into if I can make sbt-DynVer work for creating unique dev artifacts in combination with sbt-release. _(which i doubt)_
1) If 1 works, I will remove custom release-process I defined so that `sbt-release` is actually involved in whole tagging, pushing tags, and incrementing new dev version on master.

**Comment by jdegoes** [comment]
I'd say let's go ahead and merge this and do followup PRs later as necessary. @ktonga Any objection?

**Comment by jdegoes** [comment]
@prayagupd Not unexpectedly, the build broke. Please [take a look](https://travis-ci.org/scalaz/scalaz-zio/builds/412909944?utm_source=email&utm_medium=notification) when you get a chance.

**Comment by prayagupd** [comment]
Hi @jdegoes, my bad. The issue I believe is CI/CD is trying to publish version `0.1.0` to snapshot repo and getting 400 back.

I will look into it.

**Pull request merged**